protocol/testutil/app/app.go

```diff
@@ -603,6 +660,135 @@ func (tApp *TestApp) PrepareProposal() abcitypes.ResponsePrepareProposal {
+	option := cmd.GetOptionWithCustomStartCmd()
+	rootCmd := cmd.NewRootCmdWithInterceptors(
+		option,
+		// Inject the app options and logger
+		func(serverCtxPtr *server.Context) {
+			for key, value := range appOptions {
+				serverCtxPtr.Viper.Set(key, value)
+			}
+
+			// Set the test logger instance based upon AppOptions.
+			if logger, ok := appOptions[testlog.LoggerInstanceForTest]; ok {
+				serverCtxPtr.Logger = logger.(log.Logger)
+			}
+		},
+		// Override the addresses to use domain sockets to avoid port conflicts.
+		func(s string, appConfig *cmd.DydxAppConfig) (string, *cmd.DydxAppConfig) {
+			// Note that the domain sockets need to typically be ~100 bytes or fewer otherwise they will fail to be
+			// created. The actual limit is OS specific.
+			apiSocketPath := filepath.Join(validatorHomeDir, "api_socket")
+			grpcSocketPath := filepath.Join(validatorHomeDir, "grpc_socket")
+			grpcWebSocketPath := filepath.Join(validatorHomeDir, "grpc_web_socket")
+			appConfig.API.Address = fmt.Sprintf("unix://%s", apiSocketPath)
+			appConfig.GRPC.Address = fmt.Sprintf("unix://%s", grpcSocketPath)
+			appConfig.GRPCWeb.Address = fmt.Sprintf("unix://%s", grpcWebSocketPath)
+
+			// TODO(CORE-29): This disables launching the daemons since not all daemons currently shutdown as needed.
+			appConfig.API.Enable = false
+			return s, appConfig
+		},
+		// Capture the application instance.
+		func(app *app.App) *app.App {
+			appCaptor <- app
+			return app
+		},
+	)
+
+	// Specify the start-up flags.
+	// TODO(CLOB-930): Allow for these flags to be overridden.
+	rootCmd.SetArgs([]string{
+		"start",
+		// Do not start tendermint.
+		"--grpc-only",
+		"true",
+		"--home",
+		validatorHomeDir,
+		// TODO(CORE-29): Allow the daemons to be launched and cleaned-up successfully by default.
+		"--price-daemon-enabled",
+		"false",
+		"--bridge-daemon-enabled",
+		"false",
+		"--liquidation-daemon-enabled",
+		"false",
+	})
+
+	ctx := svrcmd.CreateExecuteContext(parentCtx)
+	rootCmd.PersistentFlags().String(
+		flags.FlagLogLevel,
+		tmcfg.DefaultLogLevel,
+		"The logging level (trace|debug|info|warn|error|fatal|panic)",
+	)
+	rootCmd.PersistentFlags().String(
+		flags.FlagLogFormat,
+		tmcfg.LogFormatPlain,
+		"The logging format (json|plain)",
+	)
+	executor := tmcli.PrepareBaseCmd(rootCmd, app.AppDaemonName, app.DefaultNodeHome)
+	// We need to launch the root command in a separate go routine since it only returns once the app is shutdown.
+	// So we wait for either the app to be captured representing a successful start or capture an error.
+	go func() {
+		// ExecuteContext will block and will only return if interrupted.
+		done <- executor.ExecuteContext(ctx)
+	}()
+	select {
+	case a = <-appCaptor:
+	case err = <-done:
+		// Send the error to done channel so that `Cleanup` function will not block.
+		cancelFn()
+		done <- err
+		return "", nil, nil, err
+	}
+
+	shutdownFn = func() error {
+		cancelFn()
```

Comment:
Why are we maybe calling `cancelFn` twice?

Reply:
Its a different one since the names are limited to the enclosing block.

Reply:
ICIC, go case statements do not fall through.

Reply:
that is right